### PR TITLE
TravisCI -> GHA

### PR DESCRIPTION
**Pull request by mattBrzezinski**

### CODE REVIEW

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors cancel

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by DilumAluthge** [comment]
Try this:

```yaml
on:
  push:
    branches:
      - master
      - staging
      - trying
    tags: '*'
```

**Comment by DilumAluthge** [comment]
Probably need to do `bors try-` and `bors try`

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by DilumAluthge** [comment]
We saw this issue a while back on AutoMerge, remember? GHA uses Azure, and Azure uses some of the same IPs that AWS uses.

**Comment by DilumAluthge** [comment]
It was this PR:
- "New package: AWS v1.0.0" - https://github.com/JuliaRegistries/General/pull/20506

And you implemented:
- https://github.com/JuliaCloud/AWS.jl/pull/159

**Comment by DilumAluthge** [comment]
From https://github.com/JuliaRegistries/General/pull/20506#issuecomment-683452459:
> > Or maybe GitHub/Azure is doing something funky with the IP address that you are trying to reach.
> 
> [Microsoft Azure](https://docs.microsoft.com/en-us/azure/virtual-machines/windows/instance-metadata-service) does a similar instance metadata service as AWS does, using the same IP address. This results in the `AWS.jl` package returning a `404` response instead of the expected `Timeout`.
> 
> > Is there a more robust way of checking if you are on an EC2 instance?
> 
> Yes, we do some other [checks](https://github.com/JuliaCloud/AWS.jl/blob/master/src/AWSCredentials.jl#L112) looking for various files and their contents.
> 
> > Also, I wonder what is causing the error that Fredrik posted.
> 
> It looks like I missed a small `if` check when attempting to retrieve credentials. It looks like @fredrikekre does not have an AWS Credentials on the machine they were using so the package would error out. I was able to replicate this on my machine by:
> 
>     * Wiping any AWS Environment Variables
> 
>     * Renaming my `~/.aws/config` and `~/.aws/credentials` files
> 
> 
> [JuliaCloud/AWS.jl#157](https://github.com/JuliaCloud/AWS.jl/pull/157) should allow for `AutoMerge` to pass once it goes through!

**Comment by DilumAluthge** [comment]
So my guess is this is why your tests are failing:
> Microsoft Azure does a similar instance metadata service as AWS does, using the same IP address. This results in the AWS.jl package returning a 404 response instead of the expected Timeout.

**Comment by mattBrzezinski** [comment]
> So my guess is this is why your tests are failing:
> 
> > Microsoft Azure does a similar instance metadata service as AWS does, using the same IP address. This results in the AWS.jl package returning a 404 response instead of the expected Timeout.

The tests are failing because I did not setup encrypted env vars for the credentials. :) I did this on purpose so they would die fast when testing around. Just added some in now!

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors retry

**Comment by DilumAluthge** [comment]
<details>

<summary>Here's some useful code: (click to expand)</summary>

```julia
import GitHub
import JSON
import JSON3

function branchname_to_commit(api::GitHub.GitHubAPI,
                              repo::GitHub.Repo,
                              branchname::AbstractString)
    branches, _ = GitHub.branches(repo)
    branchnames = [b.name for b in branches]
    branch = branches[findfirst(branchnames .== branchname)]
    return branch.commit
end

function status_names(api::GitHub.GitHubAPI, repo::GitHub.Repo, commit::GitHub.Commit)
    return status_names(api, repo, commit.sha)
end

function status_names(api::GitHub.GitHubAPI,
                      repo::GitHub.Repo,
                      sha::AbstractString)
    # combined_status = GitHub.status(api, GitHub.repo("JuliaRegistries/General"), "94265584ec35231ca306b40e9cf0412766936432")
    combined_status = GitHub.status(api, repo, sha)
    statuses = combined_status.statuses
    return [s.context for s in statuses]
end

function checkrun_names(api::GitHub.GitHubAPI, repo::GitHub.Repo, commit::GitHub.Commit)
    return checkrun_names(api, repo, commit.sha)
end

function checkrun_names(api::GitHub.GitHubAPI,
                        repo::GitHub.Repo,
                        sha::AbstractString)
    endpoint = "/repos/$(repo.full_name)/commits/$(sha)/check-runs"
    check_runs = GitHub.gh_get_json(
        api,
        endpoint;
        headers = Dict("Accept" => "application/vnd.github.antiope-preview+json",),
    )
    return [c["name"] for c in check_runs["check_runs"]]
end
```

</details>

**Comment by DilumAluthge** [comment]
Running that code gives me:

```julia
julia> api = GitHub.DEFAULT_API
GitHub.GitHubWebAPI(URI("https://api.github.com"))

julia> repo = GitHub.repo("JuliaCloud/AWSS3.jl");

julia> branchname = "trying"
"trying"

julia> commit = branchname_to_commit(api, repo, branchname);

julia> sha = commit.sha
"98d60375e07401242f1816b04d56734245ba3225"

julia> status_names(api, repo, sha)
Union{Nothing, String}[]

julia> checkrun_names(api, repo, sha)
7-element Vector{String}:
 "Julia 1 - ubuntu-latest - x86"
 "Julia nightly - ubuntu-latest - x64"
 "Julia 1.3 - ubuntu-latest - x64"
 "Julia 1.0.5 - ubuntu-latest - x64"
 "Julia 1 - windows-latest - x64"
 "Julia 1 - macOS-latest - x64"
 "Julia 1 - ubuntu-latest - x64"
```

**Comment by DilumAluthge** [comment]
So, if my guess is correct, your `bors.toml` file needs to look like this:
```toml
status = ["Julia 1 - ubuntu-latest - x86", "Julia nightly - ubuntu-latest - x64", "Julia 1.3 - ubuntu-latest - x64", "Julia 1.0.5 - ubuntu-latest - x64", "Julia 1 - windows-latest - x64", "Julia 1 - macOS-latest - x64", "Julia 1 - ubuntu-latest - x64"]
delete-merged-branches = true

```

**Comment by DilumAluthge** [comment]
Try that and let me know if it works.

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by DilumAluthge** [comment]
Also, if you want to allow one or more of the jobs to fail, just remove (or comment out) those particular entries from the `bors.toml` file.

**Comment by mattBrzezinski** [comment]
bors try

**Comment by DilumAluthge** [comment]
You should probably remove the Travis badge from the README and put in a GH badge.

E.g. this:
```md
![CI](https://github.com/JuliaCloud/AWSS3.jl/workflows/CI/badge.svg)
```

Which looks like this:

CI

**Comment by mattBrzezinski** [comment]
I think we have some issues with the testing now due to CI contention which wasn't an issue on Travis...

bors try

File: bors.toml [hunk 1]
```diff
@@ -0,0 +1,8 @@
+status = [
+    "Julia 1.0.5 - ubuntu-latest - x64",
+    "Julia 1.3 - ubuntu-latest - x64",
+    "Julia 1 - windows-latest - x64",
```
Comment (DilumAluthge) [inline bors.toml:4]:
```suggestion
    # "Julia 1 - windows-latest - x64",
```

**Comment by DilumAluthge** [comment]
The Windows failure is because you try to write to `/tmp` on Windows.

File: bors.toml (hunk 1 above)
Comment (mattBrzezinski) [reply bors.toml]: Ah good catch. I was basing all of this off of [JLSO.jl](https://github.com/invenia/JLSO.jl/blob/master/.github/workflows/CI.yml)'s CI setup.

I guess one day I'll look into enabling Windows CI, but today's not that day.

**Comment by DilumAluthge** [comment]
You can limit the number of jobs that are running at once: https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymax-parallel

But I don't think that's what's going on here.

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
This is the one thing I strongly dislike about bors... the amount of noise in the comments section when checking things gets a bit ridiculous.

**Comment by DilumAluthge** [comment]
Yeah I wish it would just update its first comment, instead of always posting a new comment.

**Comment by DilumAluthge** [comment]
Alright, now I do think you're running into CI contention.

Per https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymax-parallel

```yaml
strategy:
  max-parallel: 1
```

File: .github/workflows/CI.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+      - staging
+      - trying
+    tags: '*'
+jobs:
+  test:
+    name: Julia ${{ matrix.version }} - ${{ matrix.os }} - ${{ matrix.arch }}
+    runs-on: ${{ matrix.os }}
+    continue-on-error: ${{ matrix.version == 'nightly' }}
+    strategy:
+      fail-fast: false
```
Comment (DilumAluthge) [inline .github/workflows/CI.yml:16]:
```suggestion
		       fail-fast: false
max-parallel: 2
```

**Comment by DilumAluthge** [comment]
Never mind, seems to be working fine

**Comment by mattBrzezinski** [comment]
> Alright, now I do think you're running into CI contention.
> 
> Per https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#jobsjob_idstrategymax-parallel
> 
> ```yaml
> strategy:
>   max-parallel: 1
> ```

Actually I had modified the tests where this contention should not matter. Strange that bors is still waiting for the status to be reported.

**Comment by DilumAluthge** [comment]
That's because you haven't done `bors merge` yet

**Comment by DilumAluthge** [comment]
(`bors merge` is equivalent to `bors r+`)

**Comment by DilumAluthge** [comment]
Bors doesn't update the "bors" status for tries

**Comment by mattBrzezinski** [comment]
> Bors doesn't update the "bors" status for tries

Oh interesting. I thought that the `try` would update the expected check if things succeeded or failed. Kinda pointless to have this expected check icon sitting around if a `r+/merge` will just merge everything in regardless.

**Comment by DilumAluthge** [comment]
Yeah it's only useful because of the branch protection settings.

**Comment by DilumAluthge** [comment]
FYI you might run into one last issue when you try to `bors merge`. I'll keep an eye on this PR and I'll let you know if I see the issue. Basically, if this issue does occur, Bors will crash and not merge your PR, but it won't post a comment or give you any info, so you have to dig into the Bors logs to diagnose it.

**Comment by DilumAluthge** [comment]
Fingers crossed that the issue has been fixed.

**Comment by DilumAluthge** [comment]
Some Googling suggests that this issue still exists. I'm anticipating that you will run into it here, so I'll go ahead and post an explanation and the workaround.

**Comment by DilumAluthge** [comment]
# Steps to reproduce

Comment either `bors merge` or `bors r+` on this pull request. (The two commands are equivalent.)

This will trigger the following chain of events:

1. Bors force-pushes the `staging.tmp` branch to point to the same commit as the current `master`.
2. Bors merges this pull request into the `staging.tmp` branch.
3. Bors force-pushes the `staging` branch to point to `staging.tmp`
4. The GitHub Actions CI jobs run on the `staging` branch
5. All of the GitHub Actions CI jobs pass on the `staging` branch.
6. Bors detects that all of the required status checks have passed on the `staging` branch.
7. Bors attempts to fast-forward push `master` to point to the same commit as the `staging` branch

# Expected behavior

The fast-forward push succeeds. `master` now points to the same commit as the `staging` branch.

# Actual behavior

Bors crashes silently. Examination of the Bors logs reveals an error that says something like `Resource not accessible by integration`.

# Cause

Bors is a GitHub App. GitHub does not allow GitHub Apps to modify GitHub Actions workflow files on the `master` branch. Therefore, if Bors tries to fast-forward push `master` in a way that would modify a GitHub Actions workflow file, GitHub will reject the push.

# Workaround 

If this happens, you will need to manually merge this pull request.

To be clear, you can still use Bors for everything else. The only pull requests that you will need to manually merge are pull requests that modify GitHub Actions workflow files.

**Comment by DilumAluthge** [comment]
Again, I'm not 100% sure that this will happen to you.

So this is what I would recommend: when you are ready to merge this PR, comment `bors merge` or `bors r+`. And see if Bors is able to merge it. If Bors works, then nothing to worry about.

**Review by mjram0s — APPROVED** [review]

**Comment by mattBrzezinski** [comment]
Going to hold off on merging while waiting for a Slack webhooks get sorted so we can include notifications of failures.

- https://api.slack.com/messaging/webhooks
- https://github.com/marketplace/actions/notify-slack-action

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
Blanked out the secret key instead of the Access key.

**Comment by mattBrzezinski** [comment]
.....the slack notification app doesn't work on OSX...

**Comment by DilumAluthge** [comment]
This one does not use Docker: https://github.com/voxmedia/github-action-slack-notify-build

**Comment by DilumAluthge** [comment]
https://github.com/voxmedia/github-action-slack-notify-build#reporting-success-or-failure

**Comment by DilumAluthge** [comment]
This line tells us it's not a Docker action: https://github.com/voxmedia/github-action-slack-notify-build/blob/9b583ecdaddaa28bede5a056d8d8c8ff0363d66f/action.yml#L27

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try-

**Comment by mattBrzezinski** [comment]
bors try

**Comment by mattBrzezinski** [comment]
bors try-

**Pull request merged**